File: app/src/main/java/com/hk/ijournal/repository/models/AccessModel.java
package com.hk.ijournal.repository.models;

import android.text.TextUtils;

import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.Transformations;

import java.time.LocalDate;

public class AccessModel {
    //login livedata
    private MutableLiveData<String> loginUsernameLive;
    private MutableLiveData<String> loginPasscodeLive;
    private MutableLiveData<AccessValidation> loginUserValidation;
    private MutableLiveData<AccessValidation> loginPasscodeValidation;
    //register livedata
    private MutableLiveData<String> registerUsernameLive;
    private MutableLiveData<String> registerPasscodeLive;
    private MutableLiveData<LocalDate> dobLiveData;
    private MutableLiveData<AccessValidation> registerUserValidation;
    private MutableLiveData<AccessValidation> registerPasscodeValidation;

    public enum AccessStatus {
        LOGIN_SUCCESSFUL,
        USER_NOT_FOUND,
        INVALID_LOGIN,
        REGISTER_SUCCESSFULL,
        USER_ALREADY_EXISTS
    }

    public enum AccessValidation {
        USERNAME_INVALID,
        PASSCODE_INVALID,
        DOB_INVALID
    }

    DiaryUser diaryUser;

    public AccessModel() {
        diaryUser = new DiaryUser();
    }

    public MutableLiveData<String> getLoginUsernameLive() {
        if (loginUsernameLive == null)
            loginUsernameLive = new MutableLiveData<>();
        return loginUsernameLive;
    }

    public MutableLiveData<String> getLoginPasscodeLive() {
        if (loginPasscodeLive == null)
            loginPasscodeLive = new MutableLiveData<>();
        return loginPasscodeLive;
    }

    public MutableLiveData<String> getRegisterUsernameLive() {
        if (registerUsernameLive == null)
            registerUsernameLive = new MutableLiveData<>();
        return registerUsernameLive;
    }

    public MutableLiveData<String> getRegisterPasscodeLive() {
        if (registerPasscodeLive == null)
            registerPasscodeLive = new MutableLiveData<>();
        return registerPasscodeLive;
    }

    public MutableLiveData<LocalDate> getDobLiveData() {
        if (dobLiveData == null)
            dobLiveData = new MutableLiveData<>();
        return dobLiveData;
    }

    public DiaryUser getDiaryUser() {
        return diaryUser;
    }

    public void setDiaryUser(DiaryUser diaryUser) {
        this.diaryUser = diaryUser;
    }

    public MutableLiveData<AccessValidation> getLoginUserValidation() {
        if (loginUserValidation == null) {
            //transforming based on username live data
            loginUserValidation = (MutableLiveData<AccessValidation>) Transformations.map(loginUsernameLive, username -> {
                if (isUsernameInvalid(username))
                    return AccessValidation.USERNAME_INVALID;
                return null;
            });
            //setting default value
            loginUserValidation.setValue(AccessValidation.USERNAME_INVALID);
        }
        return loginUserValidation;
    }

    public MutableLiveData<AccessValidation> getLoginPasscodeValidation() {
        if (loginPasscodeValidation == null) {
            loginPasscodeValidation = (MutableLiveData<AccessValidation>) Transformations.map(loginPasscodeLive, passcode -> {
                if (isPassCodeInvalid(passcode))
                    return AccessValidation.PASSCODE_INVALID;
                return null;
            });
            //setting default value
            loginPasscodeValidation.setValue(AccessValidation.PASSCODE_INVALID);
        }
        return loginPasscodeValidation;
    }

    public MutableLiveData<AccessValidation> getRegisterUserValidation() {
        if (registerUserValidation == null) {
            registerUserValidation = (MutableLiveData<AccessValidation>) Transformations.map(registerUsernameLive, username -> {
                if (isUsernameInvalid(username))
                    return AccessValidation.USERNAME_INVALID;
                return null;
            });
            //setting default value
            registerUserValidation.setValue(AccessValidation.USERNAME_INVALID);
        }

        return registerUserValidation;
    }

    public MutableLiveData<AccessValidation> getRegisterPasscodeValidation() {
        if (registerPasscodeValidation == null) {
            registerPasscodeValidation = (MutableLiveData<AccessValidation>) Transformations.map(registerPasscodeLive, passcode -> {
                if (isPassCodeInvalid(passcode))
                    return AccessValidation.PASSCODE_INVALID;
                return null;
            });
            //setting default value
            registerPasscodeValidation.setValue(AccessValidation.PASSCODE_INVALID);
        }
        return registerPasscodeValidation;
    }

    public boolean isUsernameInvalid(String username) {
        return TextUtils.isEmpty(username);
    }

    public boolean isPassCodeInvalid(String passcode) {
        return passcode.length() < 4;
    }

    public boolean isUserValid(MutableLiveData<AccessValidation> userValidation, MutableLiveData<AccessValidation> passcodeValidation) {
        return (userValidation.getValue() == null &&
                passcodeValidation.getValue() == null);
    }

    public AccessStatus processLoginAndGetAccessStatus(DiaryUser diaryUser) {
        return signInWithUser(diaryUser);
    }

    public AccessStatus signInWithUser(DiaryUser dbUser) {
        if (dbUser == null)
            return AccessStatus.USER_NOT_FOUND;
        else if (dbUser.getPasscode().equals(diaryUser.getPasscode()))
            return AccessStatus.LOGIN_SUCCESSFUL;
        else
            return AccessStatus.INVALID_LOGIN;
    }

    public AccessStatus processRegisterAndGetAccessStatus(DiaryUser dbUser) {
        if (dbUser == null) {
            return AccessStatus.REGISTER_SUCCESSFULL;
        } else
            return AccessStatus.USER_ALREADY_EXISTS;
    }
}

File: app/src/main/java/com/hk/ijournal/repository/models/AccessModel.kt
package com.hk.ijournal.repository.models

import android.text.TextUtils
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Transformations
import java.time.LocalDate

class AccessModel {
    var diaryUser: DiaryUser

    //login livedata
    val loginUsernameLive: MutableLiveData<String> = MutableLiveData()
    val loginPasscodeLive: MutableLiveData<String> = MutableLiveData()

    //register livedata
    val registerUsernameLive: MutableLiveData<String> = MutableLiveData()
    val registerPasscodeLive: MutableLiveData<String> = MutableLiveData()
    val dobLiveData: MutableLiveData<LocalDate> = MutableLiveData()

    init {
        diaryUser = DiaryUser()
    }

    enum class AccessStatus {
        LOGIN_SUCCESSFUL, USER_NOT_FOUND, INVALID_LOGIN, REGISTER_SUCCESSFULL, USER_ALREADY_EXISTS
    }

    enum class AccessValidation {
        USERNAME_INVALID, PASSCODE_INVALID, DOB_INVALID
    }

    fun getLoginUserValidation(): MutableLiveData<AccessValidation> {
        //transforming based on username live data
        return Transformations.map(loginUsernameLive) {
            return@map if (isUsernameInvalid(it)) AccessValidation.USERNAME_INVALID
            else
                null
        } as MutableLiveData<AccessValidation>
    }

    fun getLoginPasscodeValidation(): MutableLiveData<AccessValidation> {
        //transforming based on username live data
        return Transformations.map(loginPasscodeLive) {
            return@map if (isPassCodeInvalid(it)) AccessValidation.PASSCODE_INVALID
            else
                null
        } as MutableLiveData<AccessValidation>
    }

    fun getRegisterUserValidation(): MutableLiveData<AccessValidation> {
        //transforming based on username live data
        return Transformations.map(registerUsernameLive) {
            return@map if (isUsernameInvalid(it)) AccessValidation.USERNAME_INVALID
            else
                null
        } as MutableLiveData<AccessValidation>
    }

    fun getRegisterPasscodeValidation(): MutableLiveData<AccessValidation> {
        //transforming based on username live data
        return Transformations.map(registerPasscodeLive) {
            return@map if (isPassCodeInvalid(it)) AccessValidation.PASSCODE_INVALID
            else
                null
        } as MutableLiveData<AccessValidation>
    }

    private fun isUsernameInvalid(username: String?): Boolean {
        return TextUtils.isEmpty(username)
    }

    private fun isPassCodeInvalid(passcode: String?): Boolean {
        return passcode?.let { it.length < 4 } ?: true
    }

    fun isUserValid(userValidation: LiveData<AccessValidation>, passcodeValidation: LiveData<AccessValidation>): Boolean {
        return userValidation.value != AccessValidation.USERNAME_INVALID && passcodeValidation.value != AccessValidation.PASSCODE_INVALID
    }

    fun processLoginAndGetAccessStatus(diaryUser: DiaryUser?): AccessStatus {
        return signInWithUser(diaryUser)
    }

    fun signInWithUser(dbUser: DiaryUser?): AccessStatus {
        return if (dbUser == null) AccessStatus.USER_NOT_FOUND else if (dbUser.passcode == diaryUser.passcode) AccessStatus.LOGIN_SUCCESSFUL else AccessStatus.INVALID_LOGIN
    }

    fun processRegisterAndGetAccessStatus(dbUser: DiaryUser?): AccessStatus {
        return if (dbUser == null) {
            AccessStatus.REGISTER_SUCCESSFULL
        } else AccessStatus.USER_ALREADY_EXISTS
    }
}